Updates to straggler handling functionality

ParthMandaliya opened the pull request:
This PR to resolves issue #981
 
**Summary of changes**
1. StragglerHandlingFunction: Added an interface method start_policy
2. CutoffTimeBasedStragglerHandling: start_policy method implements a timer to wait for cutoff-time and then call provided call callback method.
3. Aggregator: 
    - sendlocaltaskresults: update collaborators_done to keep track of collaborators that have finished ALL tasks
    - _straggler_cutoff_time_elapsed: call back function that is called after cutoff time has elapsed and applies the straggler policy
 
**Verification performed**

Tested various scenarios for time based cutoff straggler handling policy
All Director based workflow and Aggregator based workflow test-cases

ParthMandaliya committed:
v1.0 straggler handling added

Signed-off-by: Parth Mandaliya <parth.mandaliya.007@gmail.com>

ParthMandaliya committed:
Added start_straggler_cutoff_timer abstract function in StragglerHandlingFunction abstract class.
Renamed start_timer and __timer_expired functions to start_straggler_cutoff_timer and _straggler_cutoff_time_elapsed respectively.
Added docstring to both functions mentioned above.

Signed-off-by: Parth Mandaliya <parth.mandaliya.007@gmail.com>

ParthMandaliya committed:
Updated logs straggler handling in aggregator.py.
If one or more collaborator(s) does not even 1 task results in time, all tasks results sent by that collaborator is excluded from aggregation.

Signed-off-by: Parth Mandaliya <parth.mandaliya.007@gmail.com>

ParthMandaliya committed:
Only time based straggler handling policies require timer thread,
removing start_straggler_cutoff_timer function from parent class StragglerHandlingFunction

Signed-off-by: Parth Mandaliya <parth.mandaliya.007@gmail.com>

ParthMandaliya committed:
Find all unfinished tasks and straggler collaborators

Signed-off-by: Parth Mandaliya <parth.mandaliya.007@gmail.com>

ParthMandaliya committed:
Merge branch 'securefederatedai:develop' into straggler-handling

ParthMandaliya committed:
Review comments incorporated

Signed-off-by: Parth Mandaliya <parth.mandaliya.007@gmail.com>

ParthMandaliya committed:
Added inline comments

Signed-off-by: Parth Mandaliya <parth.mandaliya.007@gmail.com>

ParthMandaliya committed:
Changed logic to keep track of collaborators which have reported results for all tasks
Changed straggler handling logs
Added docstring for functions in all straggler handling policies

Signed-off-by: Parth Mandaliya <parth.mandaliya.007@gmail.com>

ParthMandaliya committed:
1. StragglerHandlingFunction: Added an interface method start_policy
2. CutoffTimeBasedStragglerHandling: start_policy method implements a timer to wait
for cutoff-time and then call provided call callback method
3. Aggregator:
    - sendlocaltaskresults: update collaborators_done to keep track of collaborators
	that have finished ALL tasks
    - _straggler_cutoff_time_elapsed: call back function that is called after cutoff
	time has elapsed and applies the straggler policy

Signed-off-by: Parth Mandaliya <parth.mandaliya.007@gmail.com>

ParthMandaliya committed:
Removed logger argument from start_policy &
Added logger argument to get_straggler_handling_policy in plan.py

Signed-off-by: Parth Mandaliya <parth.mandaliya.007@gmail.com>

ParthMandaliya committed:
Merge branch 'develop' into straggler-handling

ParthMandaliya committed:
If cutoff time is set to infinite, do not start the timer thread.

Signed-off-by: Parth Mandaliya <parth.mandaliya.007@gmail.com>

ParthMandaliya committed:
Resolved lint issues.

Signed-off-by: Parth Mandaliya <parth.mandaliya.007@gmail.com>

ParthMandaliya committed:
Pytest and code coverage test case failure resolved

Signed-off-by: Parth Mandaliya <parth.mandaliya.007@gmail.com>

ParthMandaliya committed:
Default logger value is set to None

Signed-off-by: Parth Mandaliya <parth.mandaliya.007@gmail.com>

ParthMandaliya committed:
Redesigned percentage based straggler policy.
minimum_reporting cannot be set 0 in any straggler policy

Signed-off-by: Parth Mandaliya <parth.mandaliya.007@gmail.com>

ParthMandaliya committed:
Code cleanup

Signed-off-by: Parth Mandaliya <parth.mandaliya.007@gmail.com>

ParthMandaliya committed:
Only collaborators_done are used for aggregation

Signed-off-by: Parth Mandaliya <parth.mandaliya.007@gmail.com>

ParthMandaliya committed:
Internal review comments incorporated
Logs updated
Logger argument removed from straggler handling policy classes

Signed-off-by: Parth Mandaliya <parth.mandaliya.007@gmail.com>

ParthMandaliya committed:
Merge branch 'securefederatedai:develop' into straggler-handling

ParthMandaliya committed:
Merge branch 'securefederatedai:develop' into straggler-handling

ParthMandaliya committed:
Resolving potential issues found during testing

Signed-off-by: Parth Mandaliya <parth.mandaliya.007@gmail.com>

ParthMandaliya committed:
Use _collaborator_task_completed method to check if all given tasks to
collaborator are completed or not

Signed-off-by: Parth Mandaliya <parth.mandaliya.007@gmail.com>

ParthMandaliya committed:
Few test cases failing issue resolved

Signed-off-by: Parth Mandaliya <parth.mandaliya.007@gmail.com>

ParthMandaliya committed:
Potential issue in aggregator based workflow tutorial resolved

Signed-off-by: Parth Mandaliya <parth.mandaliya.007@gmail.com>

ParthMandaliya committed:
Merge branch 'securefederatedai:develop' into straggler-handling

ParthMandaliya committed:
Corner case issue discovered during testing is patched.

Signed-off-by: Parth Mandaliya <parth.mandaliya.007@gmail.com>

ParthMandaliya committed:
Added reset_policy_for_round function in straggler handling policy base class.

Signed-off-by: Parth Mandaliya <parth.mandaliya.007@gmail.com>

teoparvanov commented on openfl/component/aggregator/aggregator.py:
```diff
@@ -331,11 +336,64 @@ def get_tasks(self, collaborator_name):
+        Returns:
+            None
+        """
+        # Stop timer from restarting until next round
+        self.straggler_handling_policy.is_timer_expired = True
```
It would be cleaner if the policy object itself sets `is_timer_expired` to `True` when appropriate. It doesn't seem safe to leave this responsibility with the callback function, as the developer may not be aware of it. This also breaks the encapsulation of the policy object.

ParthMandaliya replied:
Accepted, the policy object now manages the flag.
NOTE: is_timer_expired is renamed to __is_policy_applied_for_round, and the flag is managed within the class.

teoparvanov commented on openfl/component/straggler_handling_functions/straggler_handling_function.py:
```diff
@@ -5,15 +5,49 @@
+from typing import Callable
 
 
 class StragglerHandlingFunction(ABC):
```
This is no longer a pure functional interface - I'd suggest to rename it to `StragglerHandlingPolicy`

ParthMandaliya replied:
The class is renamed as suggested.
DISCUSS: In our opinion name of the file and sub-package should be renamed. If this is acceptable then this can be done in a follow-up PR.

teoparvanov replied:
Makes sense - as discussed on our sync meeting, it's ok to continue the clean-up in a separate PR.

teoparvanov commented on openfl/component/aggregator/aggregator.py:
```diff
@@ -565,6 +623,49 @@ def send_local_task_results(self, collaborator_name, round_number, task_name,
+        # NOTE: Only for %age based straggler handling policy
+        # If minimum required collaborators have reported results
+        # ignore results from all other collaborators and add them in stragglers list
+        if (
+            not hasattr(self.straggler_handling_policy, "round_start_time") and
+            self.straggler_handling_policy.straggler_cutoff_check(
+                len(self.collaborators_done), self.authorized_cols
+            )
+        ):
```
Regardless of the actual policy implementation, shouldn't we always add the lagging collaborators to the `stragglers` list (as soon as the `straggler_cutoff_check` triggers)?

ParthMandaliya replied:
The condition is now made generic for all straggler-handling policies.

teoparvanov replied:
Awesome, thanks for fixing this!

teoparvanov commented on openfl/component/straggler_handling_functions/straggler_handling_function.py:
```diff
@@ -5,15 +5,49 @@
+        raise NotImplementedError
+
+    @abstractmethod
+    def straggler_cutoff_check(self, **kwargs) -> bool:
```
Right now it's not clear what kind of inputs this method expects. Why not define it a bit more explicitly (while keeping the `**kwargs` argument for additional flexibility), as in:
```python
def straggler_cutoff_check(self, num_collaborators_done, all_collaborators, **kwargs) -> bool:
```
or even
```python
def straggler_cutoff_check(self, num_collaborators_done, num_all_collaborators, **kwargs) -> bool:
```
(for those policies, we most likely only need the list lengths, rather than the actual collaborator names)

ParthMandaliya replied:
Method definition is changed to "def straggler_cutoff_check(self, num_collaborators_done: int, num_all_collaborators: int, **kwargs) -> bool".

teoparvanov replied:
Thanks, this is much better 😊

teoparvanov reviewed (commented):
I think the code in this PR would work, but it introduces some tech debt - esp. with regards to the identified encapsulation "leaks" of the straggler policy objects. I've suggested some improvements in this first review pass, but it may be necessary to further revise the straggler policy interface.

ParthMandaliya committed:
This commit includes following changes:
1. In cutoff time based policy after cutoff time expires wait for all collaborators not just minimum required.
2. Irregardless of tasks assigned to collaborators if minimum required collaborators report resultsw in time apply straggler handling policy.

Signed-off-by: Parth Mandaliya <parth.mandaliya.007@gmail.com>

ParthMandaliya committed:
Code cleanup

Signed-off-by: Parth Mandaliya <parth.mandaliya.007@gmail.com>

ParthMandaliya committed:
Logs modified

Signed-off-by: Parth Mandaliya <parth.mandaliya.007@gmail.com>

ParthMandaliya committed:
Review comments on PR incorporated.

Signed-off-by: Parth Mandaliya <parth.mandaliya.007@gmail.com>

ParthMandaliya committed:
Log updated

Signed-off-by: Parth Mandaliya <parth.mandaliya.007@gmail.com>

ParthMandaliya committed:
Condition in straggler_cutoff_check fixed

Signed-off-by: Parth Mandaliya <parth.mandaliya.007@gmail.com>

ParthMandaliya committed:
If straggler cutoff time set to infinite only wait for minimum required collaborators to report results.

Signed-off-by: Parth Mandaliya <parth.mandaliya.007@gmail.com>

ParthMandaliya committed:
If straggler cutoff time is set to infinite wait for ALL collaborators not only for minimum_reporting collaborators

Signed-off-by: Parth Mandaliya <parth.mandaliya.007@gmail.com>

ParthMandaliya committed:
Teodor's review comments and internal review comments incorporated.
flake8 issues resovled.

Signed-off-by: Parth Mandaliya <parth.mandaliya.007@gmail.com>

ParthMandaliya committed:
Changed minimum_reporting validation.
Merged conditions in straggler_cutoff_check function in CutoffTimeBasedStragglerHandling class.

Signed-off-by: Parth Mandaliya <parth.mandaliya.007@gmail.com>

ParthMandaliya committed:
Resolved all flake8 issues.

Signed-off-by: Parth Mandaliya <parth.mandaliya.007@gmail.com>

ParthMandaliya committed:
Merge pull request #4 from ParthMandaliya/straggler-handling-v2.0

Straggler handling v2.0

ParthMandaliya committed:
Merge branch 'securefederatedai:develop' into straggler-handling

ParthMandaliya committed:
Modified single inline comment

Signed-off-by: Parth Mandaliya <parth.mandaliya.007@gmail.com>

teoparvanov commented on openfl/component/straggler_handling_functions/cutoff_time_based_straggler_handling.py:
```diff
@@ -4,30 +4,130 @@
+        ):
+            return num_all_collaborators == num_collaborators_done
+        # Something has gone, unhandled scenario, raising error.
+        raise ValueError("Unhandled scenario")
```
for debugging purposes of the "unhandled" scenario, it would be useful to also include values of the parameters from the `if/else` statements above:
```python
self.__straggler_time_expired()
self.__is_policy_applied_for_round
self.straggler_cutoff_time
```

ParthMandaliya replied:
As you suggested, I have added 3 following statements in error message:

```
self.__straggler_time_expired()
self.__is_policy_applied_for_round
self.straggler_cutoff_time
```

teoparvanov commented on openfl/component/straggler_handling_functions/cutoff_time_based_straggler_handling.py:
```diff
@@ -4,30 +4,130 @@
         **kwargs
     ):
+        if minimum_reporting <= 0:
+            raise ValueError(f"minimum_reporting cannot be {minimum_reporting}")
```
I'd add a clarification that `minimum_reporting` is expected to be a positive integer

ParthMandaliya replied:
Changed message to `"minimum_reporting must be >0"`.

teoparvanov commented on openfl/component/straggler_handling_functions/percentage_based_straggler_handling.py:
```diff
@@ -2,23 +2,51 @@
         **kwargs
     ):
+        if minimum_reporting <= 0:
+            raise ValueError(f"minimum_reporting cannot be {minimum_reporting}")
```
same here - let's explicitly state that a positive integer is expected

ParthMandaliya replied:
Changed message to `"minimum_reporting must be >0"`.

teoparvanov commented on openfl/component/aggregator/aggregator.py:
```diff
@@ -90,6 +87,10 @@ def __init__(self,
         self.log_metric_callback = log_metric_callback
 
+        self.straggler_handling_policy = (
+            straggler_handling_policy or CutoffTimeBasedStragglerHandling()
```
I suggest to explicitly set the `minimum_reporting` here to 50% of `len(authorized_cols)` when instantiating `CutoffTimeBasedStragglerHandling`. This is to account for federations with more than 2 collaborators.

ParthMandaliya replied:
Default parameter of `CutoffTimeBasedStragglerHandling()` ensure that the policy is disabled as cutoff time is set to infinite. Setting `minimum_reporting` to another value shall have no effect.

teoparvanov replied:
I understand, thanks for the clarification!

teoparvanov reviewed (approved):
LGTM, thanks for improving the encapsulation of the policy objects! It does help with code readability, but it also increases our confidence about correctness.

PS: I've left a couple of final comments from my perspective. IMHO the remaining debt can be addressed in subsequent PRs.

ParthMandaliya committed:
Merge branch 'develop' into straggler-handling

ParthMandaliya committed:
Review comments incorporated.

Signed-off-by: Parth Mandaliya <parth.mandaliya.007@gmail.com>

ishant162 committed:
Merge branch 'securefederatedai:develop' into straggler-handling

ishant162 committed:
Merge branch 'securefederatedai:develop' into straggler-handling

ishant162 committed:
Merge branch 'develop' into straggler-handling

ishant162 committed:
Lint fixes

Signed-off-by: Ishant Thakare <ishantrog752@gmail.com>

ishant162 committed:
Merge branch 'securefederatedai:develop' into straggler-handling

ishant162 committed:
Incorporated Micah's review comments and removed unused code

Signed-off-by: Ishant Thakare <ishantrog752@gmail.com>

teoparvanov commented on openfl/component/aggregator/aggregator.py:
```diff
@@ -632,7 +669,22 @@ def send_local_task_results(
+
+        self._is_collaborator_done(collaborator_name)
+
+        # If minimum required collaborators have reported results,
+        # then identify stragglers and call for early round end.
+        if self.straggler_handling_policy.straggler_cutoff_check(
+            len(self.collaborators_done), len(self.authorized_cols)
+        ):
+            self.stragglers = [
+                collab_name
+                for collab_name in self.authorized_cols
+                if collab_name not in self.collaborators_done
+            ]
+            if len(self.stragglers) != 0:
+                self.logger.warning(f"Identified straggler collaborators: {self.stragglers}")
+            self._end_of_round_check()
```
This code is an almost exact duplicate of lines 419-434. Would it make sense to extract it in a `self._end_of_round_with_stragglers_check()` method?

ishant162 replied:
Consolidated duplicate code by updating it into the `_end_of_round_with_stragglers_check()` method.

teoparvanov commented on openfl/component/aggregator/aggregator.py:
```diff
@@ -839,10 +876,11 @@ def _compute_validation_related_task_metrics(self, task_name):
+        # completed given tasks
         collaborators_for_task = []
         for c in all_collaborators_for_task:
-            if self._collaborator_task_completed(c, task_name, self.round_number):
+            if c in self.collaborators_done:
                 collaborators_for_task.append(c)
```
I'd suggest a more concise statement, using a list comprehension:
```python
collaborators_for_task = [c for c in all_collaborators_for_task if c in self.collaborators_done]
```

ishant162 replied:
Updated as per your suggestion. Thanks!

teoparvanov commented on openfl/component/straggler_handling_functions/cutoff_time_based_straggler_handling.py:
```diff
@@ -30,11 +31,100 @@ def __init__(
+                Total number of collaborators.
+
+        Returns:
+            bool
```
```python
"""
...
Returns:
        bool: True if the straggler cutoff conditions are met, False otherwise.
"""
```

ishant162 replied:
Updated. Thanks!

teoparvanov commented on openfl/component/straggler_handling_functions/cutoff_time_based_straggler_handling.py:
```diff
@@ -30,11 +31,100 @@ def __init__(
+                    "cutoff time. Applying cutoff policy and proceeding with end of round."
+                )
+                return True
+            self.logger.info(
+                "Disregarded straggler handling policy and waiting for minimum "
+                f"{self.minimum_reporting} collaborator(s) to report results."
+            )
+            return False
```
I suggest to put this case under an explicit `else:` block

ishant162 replied:
Updated. Thanks!

teoparvanov commented on tests/openfl/component/aggregator/test_aggregator.py:
```diff
@@ -171,80 +171,3 @@ def test_collaborator_task_completed_true(agg):
-
-def test_is_round_done_no_tasks(agg):
-    """Test that is_round_done returns True in the corresponded case."""
-    agg.assigner.get_all_tasks_for_round = mock.Mock(return_value=[])
-    is_round_done = agg._is_round_done()
-
-    assert is_round_done is True
-
-
-def test_is_round_done_not_done(agg):
-    """Test that is_round_done returns False in the corresponded case."""
-    round_num = 0
-    task_name = 'test_task_name'
-    col1 = 'one'
-    col2 = 'two'
-    agg.assigner.get_all_tasks_for_round = mock.Mock(return_value=[task_name])
-    agg.assigner.get_collaborators_for_task = mock.Mock(return_value=[col1, col2])
-    agg.collaborator_tasks_results = {
-        TaskResultKey(task_name, col1, round_num): 1,
-    }
-    is_round_done = agg._is_round_done()
-
-    assert is_round_done is False
-
-
-def test_is_round_done_done(agg):
-    """Test that is_round_done returns True in the corresponded case."""
-    round_num = 0
-    task_name = 'test_task_name'
-    col1 = 'one'
-    col2 = 'two'
-    agg.assigner.get_all_tasks_for_round = mock.Mock(return_value=[task_name])
-    agg.assigner.get_collaborators_for_task = mock.Mock(return_value=[col1, col2])
-    agg.collaborator_tasks_results = {
-        TaskResultKey(task_name, col1, round_num): 1,
-        TaskResultKey(task_name, col2, round_num): 1
-    }
-    is_round_done = agg._is_round_done()
-
-    assert is_round_done is True
```
Shouldn't these tests be compatible (i.e. passing) with the "happy" scenario of the `CutoffTimeBasedStragglerHandling` policy?

ishant162 replied:
Removed the unused `_is_round_done()` and `_is_task_done()` methods from the Aggregator, as they were identified as dead code. Additionally, updated the test case **tests/openfl/component/aggregator/test_aggregator.py** to remove any references to these removed methods.

teoparvanov reviewed (commented):
I have some additional questions and comments after the latest update.

ishant162 committed:
Merge branch 'securefederatedai:develop' into straggler-handling

ishant162 committed:
Merge branch 'securefederatedai:develop' into straggler-handling

ishant162 committed:
Incorporated Teo's review comments

Signed-off-by: Ishant Thakare <ishantrog752@gmail.com>

ishant162 committed:
Merge branch 'securefederatedai:develop' into straggler-handling

ishant162 committed:
Merge branch 'securefederatedai:develop' into straggler-handling

ishant162 committed:
Merge branch 'securefederatedai:develop' into straggler-handling

MasterSkepticista commented on openfl/component/aggregator/aggregator.py:
```diff
@@ -140,6 +137,10 @@ def __init__(
         self.log_metric_callback = log_metric_callback
 
+        self.straggler_handling_policy = (
+            straggler_handling_policy or CutoffTimeBasedStragglerHandling()
```
If the intention is for `CutoffTimeBasedStragglerHandling` to be default, move this to `__init__`. User should not have to dig into the code to identify the default.

ishant162 replied:
Accepted. Code Updated

ishant162 replied:
The code change was reverted due to failing Pytest and code coverage checks. Additional modifications are required and will be addressed in a separate follow-up PR

psfoley replied:
@MasterSkepticista can this be marked as resolved?

psfoley replied:
I spoke with @ishant162 about this offline. This will be resolved in a future PR.

MasterSkepticista commented on openfl/component/aggregator/aggregator.py:
```diff
@@ -177,6 +178,10 @@ def __init__(
 
         self.collaborator_task_weight = {}  # {TaskResultKey: data_size}
 
+        # maintain a list of collaborators that have completed task and
+        # reported results in a given round
+        self.collaborators_done = []
+
```
If this variable is only used by straggler policies, it should be stored in the policy object. Intent is to not spill a particular component's needs into other component's code. 

Example APIs: `self.straggler_policy.reset()` or `self.straggler_policy.mark_as_done(...)`

ishant162 replied:
It is not only used by straggler policies but also to identify collaborators who have completed all their tasks in the 
`_compute_validation_related_task_metrics()` function.

MasterSkepticista commented on openfl/component/aggregator/aggregator.py:
Class attribute docstring is out of line with the actual arguments. Can you please fix that as well?

ishant162 replied:
This change was introduced in seperate PR: https://github.com/securefederatedai/openfl/pull/1003. 
Unsure whether the changes to class attribute were intentional. Suggested to be discussed seperately as they do not seem to be related to straggler handling.

MasterSkepticista commented on openfl/component/__init__.py:
```diff
@@ -14,5 +14,5 @@
     PercentageBasedStragglerHandling,
 )
 from openfl.component.straggler_handling_functions.straggler_handling_function import (
```
Even to a developer, this naming scheme is unintuitive. How can we expect users to remember?
Please eliminate redundancy in naming modules.

```python
openfl.component.straggler_policy.CutOffPolicy
# This is self-documenting, and easy to remember than

openfl.component.straggler_handling_functions.straggler_handling_function.CutoffTimeBasedStragglerHandling
```

If this also means merging the three policy modules into one file, consider doing so. The code complexity of any policy does not justify having separate modules.

ishant162 replied:
Teo had alternate naming suggestions https://github.com/securefederatedai/openfl/pull/996#discussion_r1689369484.
Recommended that naming and other suggestions to be taken up in a seperate follow-up PR.

MasterSkepticista commented on openfl/component/straggler_handling_functions/cutoff_time_based_straggler_handling.py:
```diff
@@ -3,18 +3,19 @@
-class CutoffTimeBasedStragglerHandling(StragglerHandlingFunction):
-    """Cutoff time based Straggler Handling function."""
-
+class CutoffTimeBasedStragglerHandling(StragglerHandlingPolicy):
```
Missing docstring (explaining usage and behavior, not the arguments).

ishant162 replied:
Updated

MasterSkepticista commented on openfl/component/straggler_handling_functions/cutoff_time_based_straggler_handling.py:
```diff
@@ -30,11 +31,101 @@ def __init__(
                 collaborators that should report. Defaults to 1.
             **kwargs: Variable length argument list.
         """
+        if minimum_reporting <= 0:
+            raise ValueError("minimum_reporting must be >0")
```
Shouldn't the default behavior be to accept all reportings and remain synchronous, i.e. `inf`?

ishant162 replied:
The check ensures that the `minimum_reporting` value is greater than zero, as atleast one collaborator is required to end the round.

MasterSkepticista commented on openfl/component/straggler_handling_functions/cutoff_time_based_straggler_handling.py:
```diff
@@ -30,11 +31,101 @@ def __init__(
         self.round_start_time = round_start_time
         self.straggler_cutoff_time = straggler_cutoff_time
         self.minimum_reporting = minimum_reporting
+        self.logger = getLogger(__name__)
```
I have seen this used in many places throughout OpenFL. 
Any good reason to create per-object loggers? Can we not use the method below?

```python
import logging
logger = logging.getLogger(__name__)
...

logging.info(...)
```

ishant162 replied:
There is no specific reason to follow this approach, as both options are used in OpenFL.

MasterSkepticista commented on openfl/component/straggler_handling_functions/cutoff_time_based_straggler_handling.py:
```diff
@@ -30,11 +31,101 @@ def __init__(
+        """
+        Reset timer for the next round.
+        """
+        if hasattr(self, "timer"):
+            self.timer.cancel()
+            delattr(self, "timer")
```
I think this is a dangerous design choice to use `delattr` in classes. It is deleting attributes that other methods may call with no protection or knowledge thereof.

Why can't we use a simple `time.monotonic()` to store values in a variable and assign it to `0`, or `-1` to achieve the same goal?

[Here (internal)](https://github.com/intel-sandbox/orion-jax/blob/27fd37734b0cd98fbb56bf1fb20e6f94a9a3d071/orion/utils.py#L339) is an example stopwatch implementation for reference. Tweak as you like.

ishant162 replied:
Threading.timer is used to implement a cutoff timer that manages the CutOffTimeBasedPolicy. 
 
While trying to incorporate the feedback and suggestion to avoid deleting the timer attribute that was discussed yesterday, we realized that there is a need to delete the threading.timer object every round. Since the start method can be called only once per object (see [threading — Thread-based parallelism — Python 3.12.6 documentation](https://docs.python.org/3/library/threading.html#threading.Thread.start)), therefore there is a need to delete the old object at end of round and instantiate a new object every round
 
Further our understanding is that since the attribute is only used within the CutOffTimeBasedPolicy in start_policy() at the start of the round and reset_policy_for_round() at the end of round, therefore it is relatively safe to be used within the policy (if required for some more reason)

psfoley replied:
@ishant162 I would also recommend `time.monotonic()` to track the time elapsed - it will be far more readable for the next person who wants to touch / understand this code.

psfoley replied:
Discussed offline. `threading.Timer()` is necessary to trigger the check to determine whether the time has elapsed in the case that the final remaining collaborator in the middle of a long running task and others have stopped responding. Deletion of timer attribute is intended to free up memory (may not be necessary IMO). In 1.7 I see several opportunities for optimization here, but this code is well tested and addresses multiple issues in develop, so I'm marking this as resolved.

MasterSkepticista commented on openfl/component/straggler_handling_functions/cutoff_time_based_straggler_handling.py:
```diff
@@ -30,11 +31,101 @@ def __init__(
+            "Unhandled scenario"
+            f"{self.__straggler_time_expired()=}"
+            f"{self.straggler_cutoff_time=}"
+        )
```
This is an extremely unhelpful error message. At the very least, it should mention what was expected and what happened.

ishant162 replied:
Accepted. code updated

MasterSkepticista commented on openfl/utilities/optimizers/numpy/adam_optimizer.py:
Is this a breaking change? It seems unrelated to the PR.

ishant162 replied:
Looks like an intermediate merge. Modified as per main line

MasterSkepticista commented on openfl/component/straggler_handling_functions/straggler_handling_function.py:
```diff
@@ -7,15 +7,48 @@
 
 
-class StragglerHandlingFunction(ABC):
+class StragglerHandlingPolicy(ABC):
```
Eliminate redundancy, for a user would already know what is being imported by the time they type the module path.

Suggested: `openfl.component.straggler_policy.BasePolicy`
(+ prev comment on merging all classes to single module as it is simple logic)

ishant162 replied:
This can be handled along with https://github.com/securefederatedai/openfl/pull/996#discussion_r1751767679  recommended to be done as a seperate PR.

MasterSkepticista commented on openfl/component/straggler_handling_functions/percentage_based_straggler_handling.py:
I think percentage based cutoff is using an incorrect logic.

Percentage-based cutoff should be a specialization of time-based cutoff. 

If all collaborators take 1 second to finish each task, give or take, the policy should not go ahead with a x% cutoff "just-because" some collaborators won the race to call RPC. In this case there are no stragglers per-se.

Ideally, this policy should give a time budget to each collaborator, and then implement a percentage cutoff if the time budget is violated at each participant. What do you think?

ishant162 replied:
Our understanding is that Percentage-based policy is not a subset of time-based policy, it is rather provided as an alternative to time-based straggler handling policy. 
Due to this, suggestion of treating it as a specialization of cutoff policy cannot be considered
"Ideally, this policy should give a time budget to each collaborator, and then implement a percentage cutoff if the time budget is violated at each participant." - this seems quite similar to CutOffTimebased policy where Aggregator waits for a certain time for Collaborators to report before proceeding with early round end.

ishant162 committed:
Incorporated review comments & added mutex in aggregator for thread safety

Signed-off-by: Ishant Thakare <ishantrog752@gmail.com>

ishant162 committed:
Merge branch 'securefederatedai:develop' into straggler-handling

teoparvanov commented on openfl/component/aggregator/aggregator.py:
```diff
@@ -177,6 +178,13 @@ def __init__(
+        self.collaborators_done = []
+
+        # Initialize a mutex lock for thread safety
+        self.mutex = Lock()
```
nit - a mutex is a lower-level construct (used for the implementation of the Lock object under the hood). Therefore, I suggest to preserve the high-level abstraction here, and simply write:
```python
self.lock = Lock()
```

ishant162 replied:
Updated as per your suggestion. Thanks!

ishant162 committed:
Review comments incorporated and updated logs

Signed-off-by: Ishant Thakare <ishantrog752@gmail.com>

ishant162 committed:
Merge branch 'securefederatedai:develop' into straggler-handling

ishant162 committed:
Reverted a comment for Pytest and code coverage fix

Signed-off-by: Ishant Thakare <ishantrog752@gmail.com>
